### PR TITLE
Support Kraken and Bracken db_params in database sheet 

### PR DESCRIPTION
**Pull request by jfy133**
Closes: https://github.com/nf-core/taxprofiler/issues/233

Basically requires bracken `db_params` entries for Bracken to be a semi-colon list, with the former entry going into Kraken and the latter into Bracken.

If bracken supplied and the db_params does not include `;`, it will fail.

I will upload the different samplesheets I tested on, by using `.dumps` before and after the `map` where I parse the db_params string in `profiling.nf`

Note this will fail on first CI run due to our current samplesheet being invalid

Requires: https://github.com/nf-core/test-datasets/pull/753

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
Files used for testing (bonly == bracken only, konly == krakenonly)

[database_split_bonly.csv](https://github.com/nf-core/taxprofiler/files/10685560/database_split_bonly.csv)
[database_split_both.csv](https://github.com/nf-core/taxprofiler/files/10685561/database_split_both.csv)
[database_split_broken.csv](https://github.com/nf-core/taxprofiler/files/10685562/database_split_broken.csv)
[database_split_konly.csv](https://github.com/nf-core/taxprofiler/files/10685564/database_split_konly.csv)
[database_split_none.csv](https://github.com/nf-core/taxprofiler/files/10685565/database_split_none.csv)

File: docs/usage.md [hunk 1]
```diff
@@ -95,22 +95,24 @@ An example database sheet can look as follows, where 5 tools are being used, and
...
-bracken,db1,,/<path>/<to>/bracken/testdb-bracken.tar.gz
+bracken,db1,;-r 150,/<path>/<to>/bracken/testdb-bracken.tar.gz
 kraken2,db2,--quick,/<path>/<to>/kraken2/testdb-kraken2.tar.gz
 krakenuniq,db3,,/<path>/<to>/krakenuniq/testdb-krakenuniq.tar.gz
 centrifuge,db1,,/<path>/<to>/centrifuge/minigut_cf.tar.gz
 metaphlan3,db1,,/<path>/<to>/metaphlan3/metaphlan_database/
 motus,db_mOTU,,/<path>/<to>/motus/motus_database/
 ```
 
+For Bracken, if you wish to supply amy parameters to either the Kraken or Bracken step you **must** have a _semi-colon_ `;` list as in `db_params`. This is to allow to specify the Kraken2 parameters before, and Bracken parameters after the `;` as Bracken is a two step process. This is particularly important if you supply a Bracken database with a non-default read length parameter. If you do not have any parameters to specify, you can leave this as empty.
```
Comment (Midnighter) [inline docs/usage.md:106]:
```suggestion
For Bracken, if you wish to supply any parameters to either the Kraken or Bracken step you **must** have a _semi-colon_ `;` list as in `db_params`. This is to allow to specify the Kraken2 parameters before, and Bracken parameters after the `;` as Bracken is a two step process. This is particularly important if you supply a Bracken database with a non-default read length parameter. If you do not have any parameters to specify, you can leave this as empty.
```

File: subworkflows/local/profiling.nf [hunk 2]
```diff
@@ -120,8 +120,28 @@ workflow PROFILING {
...
+                                            db_meta_new['db_params'] = parsed_params[0]
+                                        } else if ( parsed_params.size() == 0 ) {
+                                            db_meta_new['db_params'] = ""
+                                        } else {
+                                            db_meta_new['db_params'] = parsed_params[0]
+                                        }
+
+                                    [ meta, reads, db_meta_new, db ]
+                                }
+                                .dump(tag: "ch_input_for_kraken2_after")
```
Comment (Midnighter) [inline subworkflows/local/profiling.nf:144]: Remove the dump?

```suggestion
```

File: subworkflows/local/profiling.nf [hunk 3]
```diff
@@ -160,10 +180,35 @@ workflow PROFILING {
             .filter { meta, db -> meta['tool'] == 'bracken' }
             .map { meta, db -> [meta['db_name'], meta, db] }
 
-        // Extract the database name to combine by.
+        // Combine back with the reads
         ch_input_for_bracken = ch_kraken2_output
             .map { meta, report -> [meta['db_name'], meta, report] }
             .combine(ch_bracken_databases, by: 0)
+            .dump(tag: "ch_input_for_bracken_b4")
```
Comment (Midnighter) [inline subworkflows/local/profiling.nf:187]:
```suggestion
```

File: subworkflows/local/profiling.nf [hunk 4]
```diff
@@ -160,10 +180,35 @@ workflow PROFILING {
...
+                            db_meta_new['db_params'] = ""
+                        }
+
+                    } else {
+                        db_meta_new['db_params']
+                    }
+
+                [ key, meta, reads, db_meta_new, db ]
+            }
+            .dump(tag: "ch_input_for_bracken_after")
```
Comment (Midnighter) [inline subworkflows/local/profiling.nf:211]:
```suggestion
```

**Review by Midnighter — CHANGES_REQUESTED** [review]
Just a few minor things, otherwise looks good 🙂

File: conf/modules.config [hunk 5]
```diff
@@ -359,7 +359,7 @@ process {
     }
 
     withName: BRACKEN_BRACKEN {
-        errorStrategy = 'ignore'
+        ext.args = { "${meta.db_params}" }
```
Comment (jfy133) [inline conf/modules.config:362]:
```suggestion
        errorStrategy = 'ignore'
        ext.args = { "${meta.db_params}" }
```

[the remark this answers is not in the document]
Comment (jfy133) [reply conf/modules.config:363]: Good point, I thought it was the comebine bracken output step but I remember now it was also here I think, if there were no hits?

[the remark this answers is not in the document]
Comment (Midnighter) [reply conf/modules.config:363]: Yeah, either if all is unclassified or even if only one species was detected, it fails, too, I think.

**Review by Midnighter — APPROVED** [review]

**Pull request merged**